c10/core/TensorImpl.h

```diff
@@ -228,20 +229,59 @@ struct C10_API VariableVersion {
   c10::intrusive_ptr<VersionCounter> version_counter_;
 
  public:
+  enum Disabled { DISABLED };
+  // It's okay to return true even for inference tensor which
+  // doesn't have version counter enabled.
   bool unique() const {
-    return 1 == version_counter_.use_count();
+    return this->enabled() ? 1 == version_counter_.use_count() : true;
   }
   // NOTE: As of C++11 and 14, default-constructing a std::atomic variable
   // leaves it in a persistently undefined state. See
   // https://cplusplus.github.io/LWG/issue2334.
   VariableVersion(uint32_t version = 0)
       : version_counter_(c10::make_intrusive<VersionCounter>(version)) {}
+  VariableVersion(Disabled) {}
 
-  void bump() noexcept {
-    ++version_counter_->version_;
+  bool enabled() const {
+    return version_counter_;
+  }
+
+  // Note [Inplace update inference tensor]
+  // 1. Inplace update to inference tensor is forbidden in normal mode.
+  //   For example:
+  //     inference_tensor.copy_(normal_tensor_requires_grad)
+  //   This inplace makes inference_tensor have requires_grad=True and
+  //   have a grad_fn.  This is bad because views of `inference_tensor`
+  //   created in InferenceMode won't be able to know the grad_fn since
+  //   their ViewMeta were not recorded. To match NoGradMode behavior
+  //   that "inplace update to a view created in NoGradMode raise an error",
+  //   we just ban inplace update to inference tensor since we can't tell
+  //   if an inference tensor is a view created in InferenceMode.
+  //
+  //   Note that views of `inference_tensor` created in InferenceMode has proper
+  //   ViewMeta so that they're aware of the grad_fn correctly.
+  //
+  // 2. Inplace update to inference tensor in inference tensor is doesn't bump
+  //    version counter.
+  //    * It either doesn't call bump() by skipping InplaceOrView kernel,
+  //      - e.g. inference_tensor.add_(1)
+  //    * or bump() is a no-op for inference tensor.
+  //      - e.g. inference_tensor.add_(normal_tensor)
+  void bump() {
+    bool enabled = this->enabled();
+    TORCH_CHECK(enabled || InferenceMode::is_enabled(),
+      "Inplace update to inference tensor outside InferenceMode is not allowed."
+      "You can make a clone to get a normal tensor before doing inplace update.");
+    if (enabled) {
+      ++version_counter_->version_;
```

Comment:
Nit, honestly, it's clearer if you just directly test `version_counter_` here, as it makes it clear that you are checking a possibly nullptr before dereferencing it, whereas if you indirect this through the enabled() method it is possible for enabled() to be implemented differently and then the precondition for the pointer dereference no longer holds